src/dune_pkg/opam_solver.ml

```diff
@@ -758,6 +758,26 @@ module Solver_result = struct
     }
 end
 
+let reachable pkgs ~roots =
+  let seen = ref Package_name.Set.empty in
+  let rec loop = function
+    | [] -> ()
+    | pkg :: rest ->
+      if Package_name.Set.mem !seen pkg
```

Comment:
Nitpick: Checking if a package is seen only after popping it from the stack to expand means that common packages like `ocaml` are going to be added to the stack many times. It would be more efficient to perform this check and update `seen` at the point where a package would be pushed to the stack.

Reply:
I think this does mean we might traverse the same package more than once though. For example, if `ocaml` is listed as a dependency twice.